tests/tool_use/test_parallel_tool_calls.py

```diff
@@ -0,0 +1,193 @@
+import json
+from typing import Dict, List, Optional
+
+import openai
+import pytest
+
+from .util import (MESSAGES_ASKING_FOR_PARALLEL_TOOLS,
+                   MESSAGES_WITH_PARALLEL_TOOL_RESPONSE, SEARCH_TOOL,
+                   WEATHER_TOOL)
+
+
+# test: getting the model to generate parallel tool calls (streaming/not)
+# when requested. NOTE that not all models may support this, so some exclusions
+# may be added in the future. e.g. llama 3.1 models are not designed to support
+# parallel tool calls.
+@pytest.mark.asyncio
+async def test_parallel_tool_calls(client: openai.AsyncOpenAI):
+    models = await client.models.list()
+    model_name: str = models.data[0].id
+    chat_completion = await client.chat.completions.create(
+        messages=MESSAGES_ASKING_FOR_PARALLEL_TOOLS,
+        temperature=0,
+        max_tokens=800,
```

Comment:
I _shouldn't_ need that many tokens, but, I need the model to generate a complete and correctly parsable tool call, which won't happen if it hits the `"length"` stop reason - this is a way to set a ceiling that is higher than reasonably necessary, but not so high that it can keep going and going and going and waste time if it's not going to generate a correct call. Function calls do tend to be token-intensive due to the JSON control tokens (quotes, spaces, brackets, newlines, commas, etc.) so I found that this worked well. 

I didn't spend a bunch of time counting the total generated tokens for function calls to determine precisely what this ceiling should be (and it may change with future models or different tools), but if you'd like me to zero in on a smaller value, then I'm happy to.

Reply:
I updated these so that they're all more precise values that allow for some overage (as we add additional models, formats may require more tokens) but that are much more reasonable.